### ISSUE TITLE
Support for shared folders

### ISSUE THREAD

**Issue by martensson** [ansd/lastpass-go#43]
Hi!

Just wanted to give thanks first for this library, I am porting our [Terraform Lastpass Provider](https://github.com/nrkno/terraform-provider-lastpass/tree/lastpass-go) away from lastpass-cli to use this library instead but I have encountered an issue with shared directories during my tests.

With lastpass-cli it's possible to create secrets with shared folders as part of the name, example: `Shared-Team-X/terraform-provider-lastpass test` and it will automatically split the path and use `Shared-Team-X` as the shared directory and `terraform-provider-lastpass test` as its name. It also generates a variable/field called `fullname` which concatenates both. 

I tried todo the same with this library without success, the `name` field stays the same and `share` variable is never set. 

I guess there is some differences in how they call the API, but it would be fantastic to support shared directories.

On another note is it possible for `lastpass-go` to fetch a couple of extra fields like `fullname`, `last_modified_gmt` and `last_touch`? These are not super important but nice to get full parity with the lastpass-cli client.

Either way fantastic work and would be very happy to get rid of lastpass-cli and run everything natively in Go ;)

**Comment by ansd** [comment]
Hey @martensson,

thanks a lot for using this library and for considering to port the Terraform Provider to this library. That sounds exciting :)

A couple of thoughts:
1. I pushed a commit to `master` branch which sets `last_modified_gmt` and `last_touch` when reading accounts. Right now it just returns the timestamps as strings as fetched from the LastPass servers. Could you check whether this meets your needs?
2. I'm not sure yet whether it makes a lot of sense to return `fullname` in the `Account` struct since it seems to be just `group/name` where `Group` and `Name` fields are already present in the `Account` struct. So, the client of this library could create the `fullname` itself?  I know that the CLI returns `fullname` in for example `lpass show --json {UNIQUENAME|UNIQUEID}` in addition to `group` and `name`, but if possible, I'd prefer to keep the `Account` struct simple instead of adding redundant fields. WDYT?
3. So far only reading from shared folders has been supported, but not writing. I did a spike on branch https://github.com/ansd/lastpass-go/commits/write-shared-folders which supports to `Add()` a new item to an existing shared folder. Your use case above should therefore work using this branch. It's still lacking support for `Update()` and `Delete()`. Without promising or committing myself, I plan to continue on this in a couple of weeks from now (maybe around end of October).

**Comment by martensson** [comment]
Hi @ansd!

Fantastic update, and I have made [some updates](https://github.com/nrkno/terraform-provider-lastpass/tree/lastpass-go) now to test out your new changes, and to respond your questions:

1. `last_modified_gmt` and `last_touch` works great now! I took the advantage to update the provider to convert the epoch time to a more human readable time using `time.Unix`. 
2. I agree that `fullname` is not really necessary, I mostly asked since the lastpass-cli uses it. Makes sense to keep things simple and I will drop that field as well to keep things cleaner.
3. I have tested the use of shared folders and it works fine, except changing shared folder, but that might be expected and I could use Terraforms recreate function here instead.
But I have a question regarding shared folders, I see that we use `group` to set the shared folder, but I believe `group` and `share` are different things? Looking at lastpass-cli it has two different fields for each variable at least 🤔.

Thanks again for the awesome work and if I was in the area I would buy you a beer! ;D

**Comment by ansd** [comment]
Awesome, thanks for testing it so quickly.

I just force-pushed to [write-shared-folders](https://github.com/ansd/lastpass-go/commits/write-shared-folders) branch.
This now adds support for adding, updating, and deleting items in shared folders (in addition to reading).
~~Use the new `Remove()` function when deleting an item in a shared folder.~~

>I have tested the use of shared folders and it works fine, except changing shared folder, but that might be expected and I could use Terraforms recreate function here instead.

Correct, updating an item (e.g. field `Name`, or `Password`) within a shared folder works.
Moving an item from one folder to another works as well unless source or destination folder is a shared folder.
If you want to move an item from or to a shared folder, use first function `Delete()` followed by `Add()`.
The LastPass CLI makes a special API call for that use case in [here](https://github.com/lastpass/lastpass-cli/blob/8767b5e53192ad4e72d1352db4aa9218e928cbe1/cmd-mv.c#L107-L111) and [here](https://github.com/lastpass/lastpass-cli/blob/8767b5e53192ad4e72d1352db4aa9218e928cbe1/endpoints-share.c#L303-L362) where the account ID will completely change. So, `lpass mv` is a nice helper command, but we can achieve the same by first removing and then adding the item.

While we could add the same API call (and therefore the same `lpass mv` behaviour) to this go library, we would need to keep track of the original (former) folder from the item. So for example, we could add a private field `originalShare` to every `Account` struct when reading and parsing all accounts. But then, what happens when a user wants to move an account without prior reading and parsing all accounts. Alternatively, we could fetch the blob during `Login()` and keep all data locally and whenever `Add()`, `Update()`, `Delete()` is called, the operations will happen on the local blob until some `Sync()` command is called (similarly to the LastPass CLI). Both options sound much more complex than the simple workaround of removing and then adding.

>But I have a question regarding shared folders, I see that we use group to set the shared folder, but I believe group and share are different things? Looking at lastpass-cli it has two different fields for each variable at least

Yes, that's a good point, you're right. So far, I just assumed that the 1st part of the group name is the shared folder if it starts with `Shared-`.
It seems the CLI's `fullname` has the format `Shared-folder/groupname/accountname`.
For example `lpass show --json` returns:
```json
    "fullname": "Shared-a/b/c",
    "share": "Shared-a",
    "group": "b",
    "name": "c",
```
So, does this mean we should add a new field `Share` to the [Account](https://github.com/ansd/lastpass-go/blob/efced8b59f01f796c8aa0b6541fb0f2c14bbfc50/account.go#L20-L31) struct?

Cheers 🍺😄

**Comment by martensson** [comment]
Thanks for the quick update again! Regarding the `lpass mv` functionality I think we can safely leave it if we can achieve the same result removing and then adding again. It would mean less complexity for your library, and in Terraform there is always a feature to [force recreation](https://www.terraform.io/docs/extend/schemas/schema-behaviors.html#forcenew).

And regarding the Account struct I think it is a good idea to add a new variable called `Share`, would permit us to separate it from `Group` as well 👍🏻

I believe after this I just need a bunch of testing and will hopefully be able to make the switch 🥳

Cheers mate and I hope you have a great weekend! 🍻😃

**Comment by ansd** [comment]
Thank you :)
I added the new `Share` field to the `Account` struct.
Terraform force recreation sounds great. It seems you can now set `ForceNew` only on the `Share` field.
Feel free to re-test whether everything works for you.

**Comment by martensson** [comment]
Fantastic! I have just added support for to the new `Share` variable: https://github.com/nrkno/terraform-provider-lastpass/commit/1acf4e7caccff23b72672a7c71f0af0ed9397040

I noticed that `Client.Delete()` has changed as well and now takes an `Account` instead of `ID` string. So I made some adjustments for that as well. It seems to be working pretty good!

And did I understand correctly that `ForceNew` is not necessary on the `Group` field, but only `Share`?

But this is starting to look really good and could almost at this point rewrite lastpass-cli to a native Go client without loosing any features! 😄

**Comment by ansd** [comment]
That's great! Thanks for testing. I merged, and then deleted branch `write-shared-folders` and published release `v0.3.0`.

>I noticed that Client.Delete() has changed as well and now takes an Account instead of ID string.

Yes, that's a breaking change.

>And did I understand correctly that ForceNew is not necessary on the Group field, but only Share?

Yes, in fact `ForceNew` is neither necessary on `Group` nor on `Name`. Updating these two fields works fine (independent of shared folders being involved). Updating has the advantage that the `ID` stays the same. However, updating `Share` does not work because this means the `Account` would move in / out of a shared folder. So, I think setting `ForceNew` needs only be set on field `Share`. And that's fine because the LastPass CLI has the same behaviour: The `lpass mv` command [generates a new account ID](https://github.com/lastpass/lastpass-cli/blob/8767b5e53192ad4e72d1352db4aa9218e928cbe1/endpoints-share.c#L303-L362) when moving items from / to shared folders.

>could almost at this point rewrite lastpass-cli to a native Go client without loosing any features! 

I thought about the same a while back. It would be nice to re-write the CLI in Go. The main advantage I see: it would support Windows (https://github.com/lastpass/lastpass-cli/issues/83).
There are some features not yet supported by this library: attachments (could be added), `lpass passwd` (changing master password), `lpass import`, `lpass export`, and all the `lpass share ...` subcommands etc. They are probably not important for a first version. However, a really nice feature of the existing LastPass CLI (which would be possible to implement in Go, but requires significant work) is to not enter the password on every CLI invocation and `lpass sync` to reduce API calls:
>An agent process will be spawned in the background on a first successful command, and all subsequent commands will use the agent for decryption, instead of asking a user for a password.

**Comment by martensson** [comment]
Fantastic, I have also updated the provider to use the new `v0.3.0` tag. I will start to run some more tests and see how things go, but it is already looking very good. Thanks also for clearing up the fact that is is possible to change `name` and `group` 👍🏻 

And regarding the `lpass` CLI client I also think we have enough features to support a basic Go client at least. I personally never used any of the commands you mentioned, but `lpass sync` and and credential caching would be a nice addition. Specially when working offline and on the go.

I believe this issue can be closed now, and quite happy to be one step closer to getting rid of `lpass` 🥳 

Thanks a lot again! 🙏🏻
